### ISSUE TITLE
yarn and npm repositories contain incorrect versions

### ISSUE THREAD

**Issue by fabsb** [DataTables/DataTables#955]
Both *npm* and *yarn* repositories contain (and will install) incorrect versions for *datatables.net*

`npm info datatable.net`
Output:
```
...
versions:
   [ '1.10.9',
     '1.10.10',
     '1.10.11',
     '1.10.12',
     '1.10.13',
     '1.10.14',
     '1.10.15',
     '2.1.1' ],
...
```

`yarn info datatable.net `
Output:
```
...
  versions:
   [ '1.10.9',
     '1.10.10',
     '1.10.11',
     '1.10.12',
     '1.10.13',
     '1.10.14',
     '1.10.15',
     '2.1.1' ],
...
```

Tried cleaning the cache of both yarn and npm with no improvement.

Because *yarn* and *npm* think *2.1.1* is a valid version, it gets installed when providing a version constraint like **>=1.10.9**.

**Comment by DataTables** [comment]
Thanks for letting me know about that. I must have tagged it by mistake - darn. I'll mark that version as deprecated and update here when done.

**Comment by apetitpa** [comment]
Any update on this ? I'm still facing the issue.

**Comment by DataTables** [comment]
Sorry. I'm going to get it fixed early next week.

**Comment by DataTables** [comment]
Contacted npm support and will let you know what the story is. The "version" for npm is correctly at 1.10.15, so it would be good if npm and yarn used that, but 2.1.1 is marked as deprecated and it would be good if it wasn't used at all.

**Comment by DataTables** [comment]
I've just tried creating a little dummy package:

```js
{
  "name": "xxtmp",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "datatables.net": ">=1.10.15"
  }
}
```

Running `npm install` will correctly install 1.10.15 and not 2.1.1. 2.1.1 is marked as deprecated, so it looks like npm (correctly) ignores this.

Can you tell me how you have npm configured that it is installing 2.1.1?

**Comment by fabsb** [comment]
I created this repo (https://github.com/fabsb/datatables-poc) trying to reproduce this issue.
Tried with fresh installs of *yarn* and *npm* via *brew* on my macOS Sierra.

```
$ yarn versions
yarn versions v0.27.5
{ http_parser: '2.7.0',
  node: '8.3.0',
  v8: '6.0.286.52',
  uv: '1.13.1',
  zlib: '1.2.11',
  ares: '1.10.1-DEV',
  modules: '57',
  openssl: '1.0.2l',
  icu: '59.1',
  unicode: '9.0',
  cldr: '31.0.1',
  tz: '2017b' }
Done in 0.03s.

$ npm version
{ npm: '5.3.0',
  ares: '1.10.1-DEV',
  cldr: '31.0.1',
  http_parser: '2.7.0',
  icu: '59.1',
  modules: '57',
  node: '8.3.0',
  openssl: '1.0.2l',
  tz: '2017b',
  unicode: '9.0',
  uv: '1.13.1',
  v8: '6.0.286.52',
  zlib: '1.2.11' }
```

As you can see from the *npm* branch eventually npm installs version 2.1.1. This seems to only occur after running *npm update*

With yarn there is actually a warning when installing for the first time and lockfile does not exist.

```
$ yarn install
yarn install v0.27.5
info No lockfile found.
[1/4] Resolving packages...
warning datatables.net@2.1.1: Packing error
[2/4] Fetching packages...
[3/4] Linking dependencies...
[4/4] Building fresh packages...
success Saved lockfile.
Done in 1.08s.
```

As you can see from the *yarn* branch, even after the first install ([here](https://github.com/fabsb/datatables-poc/commit/c3e7688d87dbb61709ec72021e9323d037e38970#diff-4266e65d4063ef9a1cf3e03b9a15a9d3R3)) version 2.1.1 gets installed for datatables.net.

```
$ cat node_modules/datatables.net/package.json
{
    "name": "datatables.net",
    "version": "2.1.1",
    "description": "DataTables for jQuery ",
    "files": [
        "js/jquery.dataTables.js"
    ],
    "main": "./js/jquery.dataTables.js",
    "keywords": [
        "filter",
        "sort",
        "DataTables",
        "jQuery",
        "table",
        "DataTables"
    ],
    "homepage": "https://datatables.net",
    "bugs": "https://datatables.net/forums",
    "license": "MIT",
    "author": {
        "name": "SpryMedia Ltd",
        "url": "http://datatables.net"
    },
    "dependencies": {
        "jquery": ">=1.7"
    },
    "repository": {
        "type": "git",
        "url": "https://github.com/DataTables/Dist-DataTables.git"
    }
}
```

I hope this helps.

**Comment by DataTables** [comment]
Perfect - thanks! I've created a [new issue on npm](https://github.com/npm/npm/issues/18171) which discusses this issue.

**Comment by zkat** [comment]
Deprecation does not have any effect on installability. Use `^1.10.15` instead of `>=1.10.15` as the specifier instead. All deprecation actually does is generate a warning when someone installs that version, and that's exactly what it's intended to do.

**Comment by valeryan** [comment]
I probably should have been a better community member and reported this and offered the same solution as @zkat a few weeks ago, but I have been just discarding the chunk from my lock file any time I run yarn upgrade and then running yarn to link the correct package before I compile.

**Comment by ohadlevy** [comment]
see https://github.com/yarnpkg/yarn/issues/4176 for more background of why its using the "wrong" version.

**Comment by ohadlevy** [comment]
I've proposed a solution at https://github.com/DataTables/Dist-DataTables-Bootstrap/pull/2

**Comment by valeryan** [comment]
@DataTables when you pushed out the fix for this I think this package was missed https://github.com/DataTables/Dist-DataTables-FixedHeader. It is still using >= for datatables.net.

**Comment by DataTables** [comment]
Thanks for spotting that! I'll grep them all tomorrow and make sure they are all correct.

**Comment by DataTables** [comment]
Realised what the issue is. I set it up to update the deps whenever I tag a version of the extensions (to keep their version correct as well as also resolving this issue). I've not tagged FixedHeader and a few extensions yet. I'll get that done and it should resolve itself when done. I should get that done this afternoon.

**Comment by Bruce17** [comment]
Can you please update the package `datatables`? The packages `datatables.net`, `datatables.net-bs` do not contain the current code for bootstrap v4.

**Comment by DataTables** [comment]
No, but the `datatables.net-bs4` package does :).

**Comment by Bruce17** [comment]
Good to know. A note in the Readme.md would be nice there to find what package. ;-)

**Comment by DataTables** [comment]
Fair point. It is highlighting on the [download page](//datatables.net/download), but I'll add information about the styling packages to the read me as well.
